File: crash_consistency/crash_cons_ram.go
package main

import (
   "math"
)

type RVector struct {
	capacity int32
	arr []int32
}

func Init(cap int32) *RVector {
	rv := new(RVector)
	rv.capacity = cap
	rv.arr = make([]int32, cap)
    return rv
} 

func (rv *RVector) cap() int32 {
    return rv.capacity
}

func (rv *RVector) get(i int32) int32 {
    return rv.arr[i]
}

func (rv *RVector) set(i int32, v int32) {
    rv.arr[i] = v
}

func (rv *RVector) resize(sz int32) {
    var newarr []int32 = make([]int32, sz)

    var ec int32
    if sz > rv.capacity {
        ec = rv.capacity
    } else {
        ec = sz
    }

    for i := int32(0); i < ec; i++ {
        newarr[i] = rv.arr[i]
	}
	
	rv.arr = nil
	rv.arr = newarr
	rv.capacity = sz
}   

var iterations int32 = 4000
var MAX_CAP int32 = 50000
var default_cap int32 = 2500

func main() {
	rv := Init(default_cap)

	for j := int32(0); j < iterations; j++ {
		var sum int32 = 0
		for i := int32(0); i < rv.cap(); i++ {
			var val int32
			val = int32(math.Ceil(float64(math.Sqrt(float64(rv.get(i))) + float64(((i + 1) * rv.cap()) % MAX_CAP))))
			rv.set(i, val)
			sum += val
		}
		var newcap int32 = sum % MAX_CAP + 1
		if (newcap <= 0) {
			newcap = default_cap
		}
		rv.resize(newcap)
	}

	println("done")
}

File: crash_consistency/crash_cons_ram.cpp
#include <iostream>
#include <math.h>
#include <stdint.h>
#include "../rvector.hpp"

using namespace std;

const int64_t iterations = 50000;
const int64_t MAX_CAP = 50000;
const int64_t default_cap = 2500;

int main(int argc, char **argv)
{
    RVector<int64_t> rv(default_cap);

    for (int64_t j = 0; j < iterations; j++) {
        int64_t sum = 0;
        for (int64_t i = 0; i < rv.cap(); i++) {
            int64_t val = (int64_t)(ceil(sqrt(rv.get(i)) + (((i + 1) * rv.cap()) % MAX_CAP)));
            rv.set(i, val);
            sum = sum + val;
        }
        int64_t newcap = (sum % MAX_CAP) + 1;
        if (newcap <= 0) {
            newcap = default_cap;
        }
        rv.resize(newcap);
        if (j == iterations - 1) cout << sum << endl;
    }

    return 0;
}
